src/ctl.c

```diff
@@ -3083,3 +3085,114 @@ experimental_hooks_remove_ctl(tsd_t *tsd, const size_t *mib, size_t miblen,
+ *
+ * A typical workflow would be composed of the following steps:
+ *
+ * (1) flush and disable tcache using mallctl("thread.tcache.enabled", ...)
+ * (2) initialize temporary arrays for input/output
+ * (3) call mallctl("experimental.utilization", ...)
+ * (4) defragment allocations with significant fragmentation, e.g.
+ *     for each allocation {
+ *         if it's fragmented {
+ *             free(...);
+ *             malloc(...);
+ *         }
+ *     }
+ * (5) enable tcache using mallctl("thread.tcache.enabled", ...)
+ *
+ * It's up to the application how to determine the significance of
+ * fragmentation relying on the three statistics returned.
+ *
+ * The caller needs to make sure that the input/output arrays are valid and
+ * their sizes are proper as well as matched, meaning:
+ * (a) newlen = n_pointers * sizeof(const void *)
+ * (b) *oldlenp = n_pointers * sizeof(size_t) * 3
+ * (c) n_pointers > 0
+ * Otherwise, the function immediately returns EINVAL without touching anything.
+ *
+ * In the rare case where there's no associated extent found for some pointers,
+ * rather than immediately terminating the computation and raising an error,
+ * the function simply zeros out the corresponding output fields and continues
+ * the computation until all input pointers are handled.  The motivations of
+ * such a design are as follows:
+ *
+ * (a) The function always either processes nothing or processes everything, and
+ * never leaves the output half touched and half untouched.
+ *
+ * (b) It facilitates usage needs especially common in C++.  A vast variety of
+ * C++ objects are instantiated with multiple dynamic memory allocations.  For
+ * example, std::string and std::vector typically use at least two allocations,
+ * one for the metadata and one for the actual content.  Other types may use
+ * even more allocations.  When inquiring about utilization statistics, the
+ * caller often wants to examine into all such allocations, especially internal
+ * one(s), rather than just the topmost one.  The issue comes when some
+ * implementations do certain optimizations to reduce/aggregate some internal
+ * allocations, e.g. putting short strings directly into the metadata, and such
+ * decisions are not known to the caller.  Therefore, we permit pointers to
+ * memory usages that may not be returned by previous malloc calls, and we
+ * provide the caller a convenient way to identify such cases.
+ */
+static int
+experimental_utilization_ctl(tsd_t *tsd, const size_t *mib, size_t miblen,
+    void *oldp, size_t *oldlenp, void *newp, size_t newlen) {
+	int ret = 0;
+
+	/* Requires both read and write */
+	if (oldp == NULL || oldlenp == NULL || newp == NULL || newlen == 0) {
+		ret = EINVAL;
+		goto label_return;
+	}
+
+	/*
+	 * Make sure newlen is an exact multiple of sizeof(const void *).
+	 * Rely on compiler to convert multiplication / division to shifts.
+	 */
+	const size_t len = newlen / sizeof(const void *);
```

Comment:
Let's move this to before the first branch, and then we can merge the two branches below into the first one -- note the body of the branch is the same; so cleaner to group them.